### PR TITLE
Restart development servers when `redwood.toml` is modified

### LINKED ISSUE #481: Restart development servers when redwood.toml is modified
The api server and webpack dev server should be restarted when the `redwood.toml` configuration is modified.

Related: #478

### PR DESCRIPTION
**Pull request by aggmoulik**
Fixes #481 

Expected Result: 
- API dev server restart when `redwood.toml` updated
- webpack dev server restart when `redwood.toml` updated

Completed: 
- API dev server restarted on updating of `redwood.toml`

Problems: 
- How webpack dev server is watching files?
- I tried to add in nodemon path, but it seems to be not working.

Signed-off-by: Moulik Aggarwal <qwertymoulik@gmail.com>

### CODE REVIEW

**Comment by aggmoulik** [comment]
Hey @thedavidprice @dthyresson 

I am lost in the webpack dev sever.

File: packages/web/package.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
     "prepublishOnly": "yarn cross-env NODE_ENV=production yarn build",
     "build:js": "babel src -d dist --extensions \".js,.ts,.tsx\"",
     "build:types": "ttsc --build --verbose",
-    "build:watch": "nodemon --watch src --ext \"js,ts,tsx\" --ignore dist --exec \"yarn build\"",
+    "build:watch": "nodemon --watch src redwood.toml --ext \"js,ts,tsx,toml\" --ignore dist --exec \"yarn build\"",
```
Comment (peterp) [inline packages/web/package.json:33]: This should be reverted since it doesn't do anything, there is no redwood.toml file here.

File: packages/api/src/logger/index.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import { PrismaClient } from '@prisma/client'
+/* eslint-disable @typescript-eslint/ban-ts-comment */
```
Comment (peterp) [inline packages/api/src/logger/index.ts:1]: I'm not sure about the changes in this file, they seem unrelated

File: packages/api/src/logger/index.ts [hunk 3]
```diff
@@ -350,11 +351,12 @@ interface PrismaLoggingConfig {
  */
 export const handlePrismaLogging = (config: PrismaLoggingConfig): void => {
   const logger = config.logger.child({
-    prisma: { clientVersion: config.db['_clientVersion'] },
+    prisma: { clientVersion: Prisma.prismaVersion },
```
Comment (dthyresson) [inline packages/api/src/logger/index.ts:354]: Hi @aggmoulik can you describe why you need this change and what it addresses in a different PR since this PR is intended to handle a dev server restart on toml edit not logging.

Also, I have been using v033.3 with Typescript and logging to do see the current Prisma version logged out appropriately.

Thanks!

File: packages/api-server/src/watch.ts [hunk 4]
```diff
@@ -30,8 +30,12 @@ build({ incremental: true }).then((buildResult) => {
     process.exit(0)
   })
 
+  const watchPaths = []
+  watchPaths.push(rwjsPaths.api.base)
+  watchPaths.push(rwjsPaths.redwoodConfig)
+
   chokidar
-    .watch(rwjsPaths.api.base, {
+    .watch(watchPaths, {
```
Comment (dthyresson) [inline packages/api-server/src/watch.ts:38]: Are you trying to restart the api server on a redwood.toml change?

And then your question is where to also to something similar on the dev server?

File: packages/web/package.json (hunk 1 above)
Comment (dthyresson) [reply packages/web/package.json:33]: I think what Peter is saying is that `redwood.toml` isn't in `src` but I think that you may have the `--watch` config wrong.

Have a look at the docs here: https://github.com/remy/nodemon#monitoring-multiple-directories

> Don't use unix globbing to pass multiple directories, e.g --watch ./lib/*, it won't work. You need a --watch flag per directory watched.

So maybe:

```
"build:watch": "nodemon --watch redwood.toml --watch src --ext \"js,ts,tsx,toml\" --ignore dist --exec \"yarn build\"",
```

?

**Comment by thedavidprice** [comment]
The Webpack Dev Server docs are here:
- https://webpack.js.org/configuration/dev-server/
- And the "watch" options here: https://webpack.js.org/configuration/watch/

You'll want to modify the settings here:
https://github.com/redwoodjs/redwood/blob/main/packages/core/config/webpack.development.js

**Comment by aggmoulik** [comment]
Will take care of it then.

File: packages/api-server/src/watch.ts (hunk 4 above)
Comment (aggmoulik) [reply packages/api-server/src/watch.ts:38]: Yeah, I did with the api server but need to see for the webpack dev server. I will try what David provided resources.

File: packages/api/src/logger/index.ts (hunk 3 above)
Comment (aggmoulik) [reply packages/api/src/logger/index.ts:354]: This is the problem I was facing here by the typing of PrismaClient as it was not building and I looked for the better solution then I came around with this.

Comment (dthyresson) [reply packages/api/src/logger/index.ts:354]: Sorry for not not understanding, but when you say

> the typing of PrismaClient as it was not building

can you provide an example (steps taken, behavior seen, code, screenshot) of the issue and what it was blocking or preventing you from doing and then after this fix how it addressed it (screenshot, example, etc). Thanks!

**Comment by aggmoulik** [comment]
@dthyresson @jtoar 
I have managed to make nodemon and webpack-dev-server work. Now, after changes in redwood.toml make api and web server will be restarted.

Also here are the screenshots for the Prisma Type Errors, I am getting here

logger_error

**Comment by peterp** [comment]
Hey there @aggmoulik,

I think it would be great if you used a webpack plugin rather than nodemon, something along the lines of:
https://github.com/pigcan/extra-watch-webpack-plugin

You can also grab the `redwood.toml` path from the `getConfigPath` function in `@redwoodjs/internal`

**Review by peterp — CHANGES_REQUESTED** [review]
Thanks for this PR:

It would be great if you:
1. removed the changes to unrelated files.
2. restored the order package.json
3. use `getConfigPath` instead of `getPaths`
4. Used a webpack plugin (see comment) instead of using nodemon.

**Comment by aggmoulik** [comment]
- [x] removed the changes to unrelated files.
- [x] restored the order package.json
- [x] use `getConfigPath` instead of `getPaths`
- [ ] Used a webpack plugin (see comment) instead of using nodemon.

Hey, @peterp , I have completed the 3 points but the last one seem to be not possible as I have read many blogs and none of the have this feature as well as I tried to do this programmatically before but it fell apart if I change the port of the web or api. It dosen't restart-with new port.

Now, on the part of webpack plugin, suggested plugin is old and not maintained but it will not solve any problems because of the above problem.

**Review by peterp — COMMENTED** [review]
Ah, I guess that makes sense in that it has to entirely restart when the port changes (a rare event, but not unexpected.). Pin the package and then this is good to go.

**Comment by peterp** [comment]
Holding thumbs that this will pass e2e. 🧑‍🤝‍🧑

**Review by thedavidprice — APPROVED** [review]
🚀
